src/shared/cpu-set-util.c

```diff
@@ -287,3 +295,87 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
         s = (CPUSet) {};
         return 0;
 }
+
+static int numa_policy_to_mempolicy(const NUMAPolicy *policy, unsigned long *maxnode, unsigned long **nodes) {
+        unsigned node, bits = 0, ulong_bits;
+        _cleanup_free_ unsigned long *out = NULL;
+
+        assert(policy);
+        assert(maxnode);
+        assert(nodes);
+
+        /* Take a shortcut here if policy is set to default or local since kernel requires that we pass NULL as nodemask */
+        if (IN_SET(policy->type, NUMA_POLICY_DEFAULT, NUMA_POLICY_LOCAL))
+                goto out;
+
+        bits = policy->nodes.allocated * 8;
+        ulong_bits = sizeof(unsigned long) * 8;
+
+
+        out = new0(unsigned long, DIV_ROUND_UP(policy->nodes.allocated, sizeof(unsigned long)));
+        if (!out)
+                return -ENOMEM;
+
+        /* We don't make any assumptions about internal type libc is using to store NUMA node mask.
+           Hence we need to convert the node mask to the representation expected by set_mempolicy() */
+        for (node = 0; node < bits; node++)
+                if (CPU_ISSET_S(node, policy->nodes.allocated, policy->nodes.set))
+                        out[node / ulong_bits] |= 1ul << (node % ulong_bits);
+
+out:
+        *nodes = TAKE_PTR(out);
+        *maxnode = bits + 1;
+
+        return 0;
+}
+
+int apply_numa_policy(const NUMAPolicy *policy) {
+        int mode, r;
+        _cleanup_free_ unsigned long *nodes = NULL;
+        unsigned long maxnode;
+
+        assert(policy);
+
+        if (get_mempolicy(NULL, NULL, 0, 0, 0) < 0 && errno == ENOSYS)
+                return -EOPNOTSUPP;
+
+        r = numa_policy_to_mempolicy(policy, &maxnode, &nodes);
+        if (r < 0)
+                return r;
+
+        switch (policy->type) {
+        case NUMA_POLICY_PREFER:
+                mode = MPOL_PREFERRED;
+                break;
+        case NUMA_POLICY_BIND:
+                mode = MPOL_BIND;
+                break;
+        case NUMA_POLICY_INTERLEAVE:
+                mode = MPOL_INTERLEAVE;
+                break;
+        case NUMA_POLICY_LOCAL:
+                mode = MPOL_LOCAL;
+                break;
+        case NUMA_POLICY_DEFAULT:
+                mode = MPOL_DEFAULT;
+                break;
+        default:
+                assert_not_reached("Invalid NUMA policy");
+        }
```

Comment:
so this mapping is the identity, no? why do we need a separate enum for this then, if we just map basically the same names on themselves?